README.md
# Better HTTP Handler (BHH) for node.js

## Quickstart

```js
const BHH = require("bhh-node");

var server = new BHH();
server.maxRequestBodySize = 500; //Bytes

server.register_handler(
    "/test/{test_arg}",
    async (request, response, argv)=>{
        console.log(`Received argument ${argv.test_arg}`)
        response.writeHead(
            200,
            {
                "Content-Type": "application/json"
            }
        );
        response.end(JSON.stringify({
            result: "success",
            args: argv
        }));
    }
);

server.register_handler(
    "/test",
    async (request, response, argv) => {
        try{
            var data = JSON.parse(request.body);
        } catch(e){
            server.errorHandler(request, response, {error:400, message:e.message}, 400);
            return;
        }
        response.writeHead(
            200,
            {
                "Content-Type": "application/json"
            }
        )
        response.end(JSON.stringify({
            result: "success",
            data: data
        }));
    },
    "POST"
)

console.log("Listening on port 8080");
server.listen(8080);
```

# Reference

## new BHH()
Constructor of the BHH class. No argument is needed.
## BHH.listen()
The same as [http.Server.listen()](https://nodejs.org/api/http.html#serverlisten)
## BHH.register_handler(path, handler, method?, optional_trail_slash?)
| Argument | Type | Description |
| --- | --- | --- |
| `path` | string | URI format. Using the `{}` bracket means that the section is a variable. `{}` is suggested to be used between two slashes. |
| `handler` | async function ([IncomingMessage](https://nodejs.org/api/http.html#class-httpincomingmessage), [ServerResponse](https://nodejs.org/api/http.html#class-httpserverresponse), Object) | The handler. `Object` in 3rd is the named variables sources from the URI. |
| `method` | string | Request method. Default: `GET` |
| `optional_trail_slash` | boolean | If the URI have optional `/` at the end. Default: `true` |

## Property defaultPages
| | |
| --- | --- |
| Type | `Array(string)` | 
| Default value | `["index.html", "index.htm"]` |

A list of filenames that can be served as static file when requesting a directory.
## Property errorHandler
| | |
| --- | --- |
| Type | function ([IncomingMessage](https://nodejs.org/api/http.html#class-httpincomingmessage), [ServerResponse](https://nodejs.org/api/http.html#class-httpserverresponse), Object, int) |
| Default value | `BHH.prototype._error` |

Normalized error handling function. 

Third argument is some basic information about the error.

Forth argument is the suggested HTTP status code to be sent.

When server built-in error is triggered (e.g. static file not found), this handler will be triggered like `server.errorHandler(request, response, {error: 404, message:"Not found"}, 404)`

## Property maxRequestBodySize
| | |
| --- | --- |
| Type | `number` |
| Default value | `1000` |

The maximum allowed size for request body.

## Property staticRoot
| | |
| --- | --- |
| Type | `string` |
| Default value | `""` |

Static file will be searched when no handler available for a path. `staticRoot` is a path that contains all static files.

CHANGELOG.md
# 0.0.1-alpha => 0.0.2-alpha:
* Handler can be async function now.
* Added handle duration to log output.
* Solved: No log output issue.


test.js
const bhh = require("./index.js");

const PORT = 8080;
var server = new bhh();


server.register_handler(
    "/",
    async (req, res, argv) => {
        res.writeHead(200, {
            "Content-Type": "text/plain"
        });

        res.end("This is an index.")
    }
)

server.register_handler(
    "/test",
    async (req, res, argv) => {
        res.writeHead(200, {
            "Content-Type": "application/json"
        });

        res.end(JSON.stringify({
            result: "success"
        }));
    }
);

server.register_handler(
    "/test",
    async (req, res, argv) => {
        console.log(req.body);
        try{
            req.setEncoding("utf8");
            var data = JSON.parse(req.body);
        } catch(e){
            server.errorHandler(req, res, {error: 400, message: "Invalid format"}, 400)
            return;
        }
        console.dir(data);
        res.writeHead(200, {"Content-Type": "application/json"});
        res.end(JSON.stringify({result:"success",data:data}));
    },
    "POST"
);

server.register_handler(
    "/test/{input_1}",
    async (req, res, argv)=>{
        console.dir(argv);
        res.writeHead(200, {
            "Content-Type": "application/json"
        });

        res.end(JSON.stringify({
            result: "success",
            argv: argv
        }));
    }
);

server.register_handler(
    "/test/{arg1}/{arg2}",
    async (req, res, argv) => {
        console.dir(argv);
        res.writeHead(200, {
            "Content-Type": "application/json"
        });

        res.end(JSON.stringify({
            result: "success",
            argv: argv
        }));
    }
);

server.register_handler(
    "/delay/{time}",
    async (req, res, argv) =>{
        await new Promise((resolve)=>{
            setTimeout(resolve, parseInt(argv.time));
        });
        res.writeHead(200,{
            "Content-Type": "text/plain"
        });
        res.end("DONE");
    }
)

console.log(`Opening server on port ${PORT}`)
server.listen(PORT);

index.js
const http = require("http");
const systempath = require("path");
const fs = require("fs");
const mime = require("mime-types");

class DuplicateError extends Error{}

const path_get_variable = /{([A-Za-z0-9_]*)}/g

class BHH{
    constructor(){
        this.defaultPages = [
            "index.html",
            "index.htm"
        ];
        this._handlers = {};
        this.staticRoot = "";
        this.maxRequestBodySize = 1000;
        this.errorHandler = this._error;
        this._server = http.createServer(async (req, res) => {
            var matchFound = false;
            var path_list = Object.keys(this._handlers);
            var handle_start = new Date();
            for(var i=0; i<path_list.length; i++){
            // Object.keys(this._handlers).forEach((path)=>{
                var path = path_list[i];
                var handler_obj = this._handlers[path];
                var match = handler_obj.regexp.exec(req.url);
                if (match) {
                    matchFound = true;
                    var argv = {};
                    handler_obj.variables.forEach((var_name, i)=>{
                        argv[var_name] = decodeURIComponent(match[i+1]);
                    });

                    if (handler_obj.handler.hasOwnProperty(req.method.toUpperCase())){
                        await new Promise((resolve, reject)=>{
                            // Fetching request body
                            let reqBody = '';
                            req.on("data", chunk =>{
                                reqBody += chunk;
                                if(reqBody.length > this.maxRequestBodySize){
                                    this.errorHandler(req, res, {error: 413, message:"Payload Too Large"}, 413);
                                }
                            });
                            req.on("end", async ()=>{
                                req.body = reqBody;
                                try{
                                    resolve(await handler_obj.handler[req.method.toUpperCase()](req, res, argv));
                                } catch (e) {
                                    console.dir(e);
                                    if(res.headersSent){
                                        // This section should not be like this, still looking for another solution.
                                        this.errorHandler(req, res, {error: 500, message: "Internal server error"}, 500);
                                    } else {
                                        this.errorHandler(req, res, {error: 500, message: "Internal server error"}, 500);
                                    }
                                }
                            });
                        });
                    } else {
                        // Path found but no handler
                        this.errorHandler(req, res, {error: 405, message: `${req.method} is not allowed for ${req.url}`}, 405);
                    }
                    break;
                }
            // });
            }

            if (!matchFound){
                // No matched path, checking static file.
                if (req.method.toUpperCase() == "GET"){
                    var filePath = systempath.join(this.staticRoot, req.url);
                    if(fs.existsSync(filePath)){
                        if (fs.lstatSync(filePath).isFile()){
                            res.writeHead(200, {
                                "Content-Type": mime.lookup(filePath),
                                "Content-Length": fs.statSync(filePath).size
                            });
                            res.end(fs.readFileSync(filePath));
                        } else if (fs.lstatSync(filePath).isDirectory()){
                            // Is directory, checking default page
                            var indexFound = false;
                            for(var i=0; i<this.defaultPages.length;i++){
                                let p = systempath.join(filePath, this.defaultPages[i]);
                                if (fs.existsSync(p)){
                                    if (fs.lstatSync(p).isFile){
                                        res.writeHead(200, {
                                            "Content-Type": mime.lookup(p),
                                            "Content-Length": fs.statSync(p).size
                                        });
                                        res.end(fs.readFileSync(p));
                                        indexFound = true;
                                        break;
                                    }
                                }
                            }
                            if (!indexFound) {
                                this.errorHandler(req, res, {error: 404, message: "Not Found"}, 404);
                            }
                        }
                    } else {
                        // Nothing found
                        this.errorHandler(req, res, {error: 404, message: "Not Found"}, 404);
                    }
                }
            }
            var handle_end = new Date();
            console.log(`${handle_end.toISOString()} ${(handle_end.getTime() - handle_start.getTime())}ms ${req.socket.remoteAddress} ${req.method} ${req.url} - ${res.statusCode}`)
        });
    }

    listen(...args){
        // Pass listen argument to builtin http server
        this._server.listen.apply(this._server, args);
    }

    _error(req, res, error, statusCode=400){
        res.writeHead(statusCode, {"Content-Type": "application/json"});
        res.end(JSON.stringify(error));
    }

    register_handler(path, handler, method="GET", optional_trail_slash=true){
        method = method.toUpperCase();
        if(! this._handlers.hasOwnProperty(path)){
            var v = [];
            var match;

            // node seems not having the ability to return all matched group, looping through all matches
            while(match = path_get_variable.exec(path)){
                v.push(match[1]);
            }

            this._handlers[path] = {
                handler: {},
                regexp: new RegExp(`^${path.replace(/\{([^\/\}]*)\}/g, "([^/]*)")}${optional_trail_slash?"/?":""}$`),
                variables: v
            };
        }
        if (this._handlers[path].handler.hasOwnProperty(method)){
            throw new DuplicateError(`${method} for "${path}" has already existed.`);
        }
        this._handlers[path].handler[method] = handler;
    }


}

module.exports = BHH;